crmprtd/download_cache_process.py

```diff
@@ -0,0 +1,371 @@
+        return f"~/{network_name}/logs/{tag}_mof_json.log"
+    return f"~/{network_name}/logs/{tag}_{network_name}_json.log"
+
+
+def cache_filename(
+    timestamp: datetime.datetime = datetime.datetime.now(),
+    timestamp_format: str = "%Y-%m-%dT%H:%M:%S",
+    network_name: str = None,
+    tag: str = None,
+    frequency: str = None,
+    province: str = None,
+    **_,
+) -> str:
+    """Return cache filename (filepath). It depends on several parameters, starting
+    with the network. Its pattern is similar, but not identical, to that of the log
+    filename.
+
+    :param timestamp: Timestamp to include in filename.
+    :param timestamp_format: Format for converting timestamp to string.
+    :param network_name: Network name to include in filename.
+    :param tag: Tag to include in filename.
+    :param frequency: Download frequency to include in filename (network ec only).
+    :param province: Province code (2 letters) to include in filename (network ec only).
+    :param _: Ignored additional kw args.
+    :return: Filename.
+    """
+    check_network_name(network_name)
+    ts = timestamp.strftime(timestamp_format)
+    if network_name in ("ec",):
+        return (
+            f"~/{network_name}/download/{tag}_{frequency}_{province.lower()}_{ts}.xml"
+        )
+    if network_name in (
+        "bc_env_snow",
+        "bc_forestry",
+        "bc_tran",
+        "dfo_ccg_lighthouse",
+        "nt_forestry",
+        "nt_water",
+        "yt_gov",
+        "yt_water",
+        "yt_avalanche",
+        "yt_firewx",
+    ):
+        return f"~/{network_name}/cache/{tag}_{network_name}_{ts}.xml"
+    if network_name in ("wmb",):
+        return f"~/{network_name}/download/{tag}_{network_name}_{ts}.txt"
+    return f"~/{network_name}/cache/{tag}_{network_name}_{ts}.txt"
+
+
+def log_args(**kwargs):
+    """Return logging args. Only the log filename depends on the arguments."""
+    return [
+        "-L",
+        "~/logging.yaml",
+        "--log_filename",
+        log_filename(**kwargs),
+    ]
+
+
+def to_utc(d: datetime.datetime, tz_string: str = "Canada/Pacific"):
+    if d.tzinfo is not None and d.tzinfo.utcoffset(d) is not None:
+        # d is not localized. Make it so.
+        tz = pytz.timezone(tz_string)
+        d = tz.localize(d)
+    return d.astimezone(pytz.utc)
+
+
+def download_args(
+    network_name: str = None,
+    frequency: str = None,
+    province: str = None,
+    time: datetime.datetime = None,  # TODO: use now()?
+    start_time: datetime.datetime = None,
+    **_,
+):
+    """Return command-line args for download phase. They depend on the network and
+    other arguments.
+
+    :param time:
+    :param start_time:
+    :param network_name: Network name.
+    :param _: Remainder args. Passed through.
+    :return:
+    """
+    check_network_name(network_name)
+
+    common_args = f"-N {network_name} ".split()
+    net_args = None
+
+    # Set net_args per network.
```

Comment:
I think that all of these need to be documented in the online help somewhere. *Especially* the auth arguments. (By necessity,) There are a lot of implied requirements that someone could easily overlook if they went to run this on a fresh system (or, say, a docker image in a K8s cron job).

Reply:
Yes. I think a "dry-run" flag, which prints rather than executes the commands, would be useful. It's straightforward. I'll try it and you can see what you think.

Reply:
@jameshiebert , here's some sample outputs of the dry-run option for `crmprtd_dcp`. I'm intending this to fill in for the rather complicated (and needing ongoing maintenance) documentation we want.

- Some have the processing step and others do not, distinguished by whether the `-c` (connection string) option is specified. 
- I edited out the `-u` argument value in case it is secret. Is it?
- Comments on format? These lines could be fed straight into a Linux command line and they would work. But maybe not so readable?

```
(crmprtd) rglover@pcic-3002:~/code/crmprtd$ crmprtd_dcp --dry_run -N moti -T tagg
crmprtd_download -N moti -u <secret?> --auth_fname ~/.rtd_auth.yaml --auth_key=moti -L ~/logging.yaml --log_filename ~/moti/logs/tagg_moti_json.log > ~/moti/cache/tagg_moti_2023-01-19T16:53:27.txt
(crmprtd) rglover@pcic-3002:~/code/crmprtd$ crmprtd_dcp --dry_run -N ec -F daily -T tag -c DSN
crmprtd_download -N ec -p bc -F daily -L ~/logging.yaml --log_filename ~/ec/logs/tag_bc_daily_json.log | tee ~/ec/cache/tag_daily_bc_2023-01-19T16:54:04.xml | crmprtd_process -N ec -c DSN -L ~/logging.yaml --log_filename ~/ec/logs/tag_bc_daily_json.log
crmprtd_download -N ec -p yt -F daily -L ~/logging.yaml --log_filename ~/ec/logs/tag_yt_daily_json.log | tee ~/ec/cache/tag_daily_yt_2023-01-19T16:54:04.xml | crmprtd_process -N ec -c DSN -L ~/logging.yaml --log_filename ~/ec/logs/tag_yt_daily_json.log
(crmprtd) rglover@pcic-3002:~/code/crmprtd$ crmprtd_dcp --dry_run -N ytnt -T tag
crmprtd_download -N nt_forestry -d "2023/01/19 23:00:00" -L ~/logging.yaml --log_filename ~/nt_forestry/logs/tag_nt_forestry_json.log > ~/nt_forestry/cache/tag_nt_forestry_2023-01-19T15:54:56.xml
crmprtd_download -N nt_water -d "2023/01/19 23:00:00" -L ~/logging.yaml --log_filename ~/nt_water/logs/tag_nt_water_json.log > ~/nt_water/cache/tag_nt_water_2023-01-19T15:54:56.xml
crmprtd_download -N yt_gov -d "2023/01/19 23:00:00" -L ~/logging.yaml --log_filename ~/yt_gov/logs/tag_yt_gov_json.log > ~/yt_gov/cache/tag_yt_gov_2023-01-19T15:54:56.xml
crmprtd_download -N yt_water -d "2023/01/19 23:00:00" -L ~/logging.yaml --log_filename ~/yt_water/logs/tag_yt_water_json.log > ~/yt_water/cache/tag_yt_water_2023-01-19T15:54:56.xml
crmprtd_download -N yt_avalanche -d "2023/01/19 23:00:00" -L ~/logging.yaml --log_filename ~/yt_avalanche/logs/tag_yt_avalanche_json.log > ~/yt_avalanche/cache/tag_yt_avalanche_2023-01-19T15:54:56.xml
crmprtd_download -N yt_firewx -d "2023/01/19 23:00:00" -L ~/logging.yaml --log_filename ~/yt_firewx/logs/tag_yt_firewx_json.log > ~/yt_firewx/cache/tag_yt_firewx_2023-01-19T15:54:56.xml
```

Reply:
I think this works well, and it is self-maintaining.